### ISSUE TITLE
[Proposal] Add await lock block

### ISSUE THREAD

**Issue by rev3r** [dotnet/csharplang#2612]
# Background

We now have `await foreach` and `await using` blocks. They are possible because of the extremely cheap `ValueTask` struct. But we still lack `await lock` block which would allow not blocking threads that are waiting for lock to be released. Instead they would be free to do other work.

# Proposal

```csharp
await lock (_obj)
{
    //code that needs to be synchronized
}
```

Can be combined with https://github.com/dotnet/csharplang/issues/2613.

# Similar issues

Other issues discusses allowing `await` inside the `lock` block. This issue discusses allowing `await` on the `lock` itself.

**Comment by yaakov-h** [comment]
If you only have `await lock` without `(obj)`, how would you determine which blocks can be run concurrently and which locks cannot?

**Comment by rev3r** [comment]
That is really good point. I'll update issue accordingly 😉

**Comment by HaloFour** [comment]
I don't see how you can mix the two safely.  Locking is thread-specific and async is not.  The thread that would eventually acquire that lock would be a different thread than the thread that requested the lock.  This would be particularly bad when it comes to reentrancy or multiple locks as you'd be guaranteed to eventually run into a deadlock.

**Comment by rev3r** [comment]
How this is different from `SemapthoreSlim WaitAsync`? That was async too.

**Comment by HaloFour** [comment]
@rev3r

`SemaphoreSlim` doesn't enforce thread identity and has no concept of reentrancy.

`Monitor`, on which `lock` is built, does both and needs to track the thread that has acquired the lock.

**Comment by HaloFour** [comment]
I've updated my comment to make it clear that I was talking about `lock`.  That was an omission on my part and I apologize for the confusion.  There are locking classes in the BCL or other projects which can work with async, but they generally work a little differently.  I'm not aware of any async locks that support reentrancy, for example.

**Comment by rev3r** [comment]
If we could create new types to support `await foreach` and `await using` concepts, we could create new `ValueTask` type that could be used inside `await lock`. I'm not really fan of `Monitor` anyway.

Async locks would allow not blocking (which is very bad), and not spinning (which wastes CPU cycles).

**Comment by mikedn** [comment]
Pff, what's up with this stream of lock related suggestions? Just couple of hours ago a different one appeared that also involves `lock` and async, though in a different manner - #2611 

> Async locks would allow not blocking (which is very bad), and not spinning (which wastes CPU cycles).

What exactly are you doing inside a lock that causes blocking for significant periods of time? The point of a lock is to synchronize access to some shared data and that shared data is usually in memory data structure. It's not like you're using a lock to synchronize access to a network resources.

**Comment by rev3r** [comment]
@mikedn 3 of those are mine, the one before is just a coincidence and about different aspect :wink:

This idea come from `SemaphoreSlim WaitAsync` combining with `ValueTask`. Async lock already exist and it's being used, but `ValueTask` would allow avoiding allocations.

**Comment by mikedn** [comment]
> This idea come from SemaphoreSlim WaitAsync combining with ValueTask. Async lock already exist and it's being used, but ValueTask would allow avoiding allocations.

Yes, but why is language support needed for this?

**Comment by rev3r** [comment]
@mikedn To align it with other new shine `await` blocks. It's really similar and it could be expressed nicely in language itself.

**Comment by mikedn** [comment]
The language doesn't get extended just to align with other shiny new things. It is extended because there's a practical need for that. Any extensions around `lock` strike me as very dubious because `lock` is not widely used. Normally it's only needed here and there and an async variant should be even less common.

Besides, `lock` is arguably a design error of the original language. C# decided to be short sighted and followed on the footsteps of Java which was even more short sighted and failed to acknowledge that not everything can be managed by GC. The result was the creation of this special case of lifetime management that better designed languages do not have and do not need (e.g. there's no `lock` statement in C++ and nobody serious asks for one because it is not needed).

**Comment by rev3r** [comment]
@mikedn So how to assure thread-safe double assignment without any kind of lock (eg. linked list)? You need `lock` for those corner cases.

So why not allow async version to not block threads that are waiting. It's particularly useful in the server with many threads accessing shared mutable resources. You don't want to block them, they could handle another request instead.

**Comment by mikedn** [comment]
> So how to assure thread-safe double assignment without any kind of lock (eg. linked list)? You need lock for those corner cases.

I didn't say that `lock` isn't needed, I said that it's not commonly needed. Not commonly enough to warrant language changes. For example, WPF repo has some 480,000 lines of code and only 1300 contain the work `lock`. That's not even 1%. And from a cursory look it seems that some of locks could have perhaps been replaced by concurrent collections.

> So why not allow async version to not block threads that are waiting. It's particularly useful in the server with many threads accessing shared mutable resources. You don't want to block them, they could handle another request instead.

Nothing's stopping you from using `SemaphoreSlim.WaitAsync` in situation where it makes sense to do so. Do you really think it would be useful to use `await lock` to, say, access a `Dictionary<K, V>` in a thread safe manner? It's extremely unlikely that anything good will come out of that.

**Comment by rev3r** [comment]
@mikedn How about server apps? UI apps are heavily single threaded, so synchronisation is lesser problem.

I honestly think that it would great, if we had `ValueTask` returning concurrent collections. Non-blocking with (almost) no additional cost ❤️

**Comment by CyrusNajmabadi** [comment]
> I honestly think that it would great, if we had ValueTask returning concurrent collections. Simple, non-blocking, with (almost) no additional cost ❤️

That is unrelated to the language.  If you'd like that, you can go ask CoreFx for it.  

As someone said above: 

> The language doesn't get extended just to align with other shiny new things. It is extended because there's a practical need for that.

**Comment by mikedn** [comment]
> How about server apps? UI apps are heavily single threaded, so synchronisation is lesser problem.

While UI apps are typically single threaded WPF itself tries to be thread safe where it is needed. The number of `lock` statements I see in it is quite high compared to other codebases I've seen. And BTW, it looks like I got the line count wrong and WPF actually has over one million lines of C# code - 1,664,636. 1300 out of that is 0.07%.

Any reasonable server app will strive to have as few locks as possible because locks hurt scalability. In general, read/write sharing hurts because it implies additional memory traffic. `await lock` won't really help with that, if there's high lock contention and threads regurarely block waiting on locks, `await` would at best reduce the number of blocked threads, it won't make things progress faster. In fact it's possible that it will make things slower due to the overhead `await` adds.

> I honestly think that it would great, if we had ValueTask returning concurrent collections. Simple, non-blocking, with (almost) no additional cost ❤️

Perhaps that would be a good suggestion for the corefx repository where concurrent collections live.

**Comment by rev3r** [comment]
@mikedn Only standard, synchronous `lock` hurts scalability, because it blocks. And could make `ThreadPool` create more, unnecessary threads as a result.

With `ValueTask` overhead is extremely low.

>That is unrelated to the language. If you'd like that, you can go ask CoreFx for it.

>Perhaps that would be a good suggestion for the corefx repository where concurrent collections live.

That's why I only mention it, and as a replay. I know that's not the repo for that.

**Comment by CyrusNajmabadi** [comment]
> @mikedn To align it with other new shine await blocks. It's really similar and it could be expressed nicely in language itself.

To be clear, this is a non-reason for teh LDM.  Features are done because of high value.     `await lock` has practically no value.  For things like `await foreach` the value is in allowing the stream elements to be asynchronously produced.  For `await dispose` it's to allow actual disposal to be async.  These have been two big pain points that we've gotten lots of feedback on.  No such feedback exists for `lock`.  The scenarios aren't there.  And the language isn't going to do new features that there isn't a realistic audience for that would greatly benefit from it.

**Comment by mikedn** [comment]
> Only standard, synchronous lock hurts scalability, because it blocks. And could make ThreadPool create more, unnecessary threads as a result.

The number of threads is only one aspect of scalability. A critical section inherently hurts scalability because it serializes execution. If a critical section requires 1 second to execute and 100 threads have to execute it then the application will require 100 seconds to finish. It may require 1 or 100 threads dependening on how locking is done but one way or another it takes 100 seconds to complete.

That's why you want to keep the number of critical sections and their execution time as low as possible. And normally it is pretty low since typical critical section involve updating some in memory data structures.

> With ValueTask overhead is extremely low.

I don't think you understand what ValueTask does. In the case of an async lock it would be useful when there's little to no contention on the lock and the async lock is acquired synchronously. If async blocking is required then a normal task needs to be allocated and the normal async machinery gets involved.

So, if `ValueTask` is useful, it is because there's no contention and if there's no contention then you may as well use the standard `lock`.

**Comment by rev3r** [comment]
> If async blocking is required then a normal task needs to be allocated

Not when using `IValueTaskSource`. The cost is one (first) allocation and that's it.

**Comment by mikedn** [comment]
> Not when using IValueTaskSource. The cost is one (first) allocation of the continuation and that's it.

OK but the async machinery is still involved. That ain't that cheap, at least for typical `lock` scenarios.

> From now on every thread is not blocked & is not allocating. Ideal for server apps.

Do you have any evidence that the current `lock` is not ideal for server apps? So far you haven't offered any practical scenario that would justify the need for an async `lock`. Again, keep in mind that critical sections are supposed to be relatively rare and execute fast. If they aren't, the number of blocked threads is likely to be the least of your concerns.

**Comment by rev3r** [comment]
@mikedn Creating prototype with proper `IValueTaskSource` implementation is beyond me. That's why I argue only on theoretical basis.

It's just make sense to me to have `ValueTask` version of `lock` block. We have all building blocks necessary.

And for sever apps, ASP.NET Core team avoids lock & blocking, and prefer asynchronous code everywhere. `await lock` is perfect middle ground - it's easy to use and fast enough.

**Comment by mikedn** [comment]
> That's why I argue only on theoretical basis.

I don't see this theoretical basis. Normally people look at one or more application, determine if there are problems (e.g. lock contention that leads to high number of blocked threads) and investigate that can be done (replace locks with lock-free approaches, use different kind of locks - reader/writer locks perhaps, minimize the amount of time spent inside critical sections etc.). Simply saying "let's do X because it seems shiny" is more in the realm of speculation, especially when the existing and well known "theory" simply says that "long running critical sections are bad".

> And for server apps, ASP.NET Core team avoids lock & blocking, and prefer asynchronous code everywhere. 

ASP.NET's main use of async is for I/O. In general, I/O is the poster child for async and async I/O has existed for a very long time in various forms. The reason is that I/O can take lots of time (hundreds of milliseconds and even seconds) and typically doesn't need a thread - obviously you don't need a thread to wait for a remote server to send you data).

> await lock fits there perfectly.

There's simply no evidence that typical locks need to be asynchronous.

**Comment by YairHalberstadt** [comment]
Closing as duplicate of https://github.com/dotnet/csharplang/discussions/2611
